qualtran/bloqs/chemistry/sparse/prepare.py

```diff
@@ -77,39 +80,84 @@ class PrepareSparse(Bloq):
+            SelectionRegister(
+                "q",
+                bitsize=(self.num_spin_orb // 2 - 1).bit_length(),
+                iteration_length=self.num_spin_orb // 2,
+            ),
+            SelectionRegister(
+                "r",
+                bitsize=(self.num_spin_orb // 2 - 1).bit_length(),
+                iteration_length=self.num_spin_orb // 2,
+            ),
+            SelectionRegister(
+                "s",
+                bitsize=(self.num_spin_orb // 2 - 1).bit_length(),
+                iteration_length=self.num_spin_orb // 2,
+            ),
+            SelectionRegister("sigma", bitsize=self.num_bits_state_prep),
+            SelectionRegister("alpha", bitsize=1),
+            SelectionRegister("beta", bitsize=1),
+            SelectionRegister("rot_aa", bitsize=1),
+            SelectionRegister("swap_pq", bitsize=1),
+            SelectionRegister("swap_rs", bitsize=1),
+            SelectionRegister("swap_pqrs", bitsize=1),
+            Register("flag_1b", bitsize=1),
+        )
+
+    @cached_property
+    def junk_registers(self) -> Tuple[SelectionRegister, ...]:
+        return (
+            Register('alt_pqrs', bitsize=(self.num_spin_orb // 2 - 1).bit_length(), shape=(4,)),
+            Register('theta', bitsize=1, shape=(2,)),
+            Register('keep', bitsize=self.num_bits_state_prep),
+            Register("less_than", bitsize=1),
+            Register("alt_flag_1b", bitsize=1),
         )
 
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         num_bits_spat = (self.num_spin_orb // 2 - 1).bit_length()
+        if self.qroam_block_size is None:
+            target_bitsizes = (
+                (num_bits_spat,) * 4
+                + (1,) * 2
+                + (num_bits_spat,) * 4
+                + (1,) * 2
+                + (self.num_bits_state_prep,)
+            )
+            block_size = 2 ** find_optimal_log_block_size(self.num_non_zero, sum(target_bitsizes))
+        else:
+            block_size = self.qroam_block_size
         if self.adjoint:
-            num_toff_qrom = int(np.ceil(self.num_non_zero / self.k)) + self.k  # A15
+            num_toff_qrom = int(np.ceil(self.num_non_zero / block_size)) + block_size  # A15
         else:
             output_size = self.num_bits_state_prep + 8 * num_bits_spat + 4
-            num_toff_qrom = int(np.ceil(self.num_non_zero / self.k)) + output_size * (
-                self.k - 1
+            num_toff_qrom = int(np.ceil(self.num_non_zero / block_size)) + output_size * (
+                block_size - 1
             )  # A14
         qrom_cost = (Toffoli(), num_toff_qrom)
         if self.adjoint:
             return {
                 (PrepareUniformSuperposition(self.num_non_zero, self.num_bits_rot_aa), 1),
                 qrom_cost,
             }
-        swap_cost_state_prep = (CSwapApprox(num_bits_spat), 4 + 4)  # 2. pg 39
+        swap_cost_state_prep = (CSwap(num_bits_spat), 4 + 4)  # 2. pg 39
```

Comment:
is this correct?

Reply:
The Approx -> regular swap? Yes (at least the approx cswap isn't referenced ever in the chemistry algorithms) so I think this is more appropriate.